### ISSUE TITLE
Missing ARG expansion for ENTRYPOINT / CMD inside of shell quotes

### ISSUE THREAD

**Issue by cirego** [earthly/earthly#1846]
I ran into an issue with a container no longer behaves the same after upgrading Earthly. I have the following example spec:

```
example:
    FROM ubuntu:18.04

    ARG --required FOO
    ENTRYPOINT ["echo", "${FOO}"]
    SAVE IMAGE example/example
```

If I build and inspect this container by running `earthly +example --FOO=bar && docker inspect example/example --format "{{.Config.Entrypoint}}"`, I see the following:

```
[echo ${FOO}]
```

This is not what I expected. Instead, I expected to see something like:

```
[echo bar]
```

I can workaround this by using the shell-form of `ENTRYPOINT`:

```
example:
    FROM ubuntu:18.04

    ARG --required FOO
    ENTRYPOINT echo ${FOO}
    SAVE IMAGE example/example
```

And now I get something that resembles what I want:

```
[/bin/sh -c echo bar]
```

Here is the output of `earthly --version`:

```
[zen] earthly --version
earthly version v0.6.13 66ae40bb5b03769635a8a0bd0d5976dfc244514c linux/amd64; Ubuntu 21.10
```

I don't recall on which version of Earthly this used to work. It seems like this may also be related to #1744.

**Comment by dfrkp** [comment]
This is still an issue in `earthly version v0.6.28`. The shell form is not a great workaround though :( Would be great to see this fixed :)
-- 
Using this opportunity though to call out the great work you guys are doing with Earthly 👍

**Comment by oderwat** [comment]
I want to use it to make the Eartfile easier to modify by setting some global. But it does not work with the entry point and I can't use shell form.

```
    ENTRYPOINT ["/app/$APP_NAME"]
    SAVE IMAGE $APP_NAME:latest
```

**Comment by nelsam** [comment]
FWIW, we're [consistent with docker](https://docs.docker.com/engine/reference/builder/#run) here. The exec form does not do variable expansion.

> Unlike the shell form, the exec form does not invoke a command shell. This means that normal shell processing does not happen. For example, RUN [ "echo", "$HOME" ] will not do variable substitution on $HOME. If you want shell processing then either use the shell form or execute a shell directly, for example: RUN [ "sh", "-c", "echo $HOME" ]. When using the exec form and executing a shell directly, as in the case for the shell form, it is the shell that is doing the environment variable expansion, not docker.

If we change this behavior, it makes us less docker-compatible - i.e. it makes fewer commands work as expected when copied/pasted from a `Dockerfile` to an `Earthfile`.

**Comment by oderwat** [comment]
I think it would be nice to add it optionally. Maybe adding something like `ENTRYPOINT --args ["bla $ME"]`. Although I doubt that real-world code would break if $  or even just ${} could get expanded.

**Comment by nelsam** [comment]
Yeah, `ENTRYPOINT --args` might be okay. We could also do something like:

> In the exec form, variables will only be expanded outside of quotes. If a variable is next to a string, the variable and string will be concatenated, e.g. `["/usr/bin/"$app, "someArg"]`

^^^ Currently, docker will fail to parse that and fall back on the shell form, resulting in the error `/bin/sh: [/usr/bin/,: not found`. I think it might be reasonable for us to expand variables outside of quotes, while leaving everything inside of quotes alone.

---

Also, just to clarify why it would be dangerous to support variable expansion in the exec form: in the shell form, we add `ARG`s as environment variables to the shell when `ENTRYPOINT` is run. This means that when an inlined shell script shadows an earthly variable, it correctly gets shadowed in the shell's context - so everything works as you'd expect. So for example:

```Earthfile
ARG PORT
ENTRYPOINT export PORT=8080; echo "starting on port $PORT"; someBinary
```
^^^ That will use the value `8080` for `$PORT`, both in the `echo` statement and in `someBinary`.

In the exec form, there is no shell. We have no way to rely on a shell to perform variable interpolation. So we would have to parse the strings ourselves and perform custom variable replacement. So for example:

```Earthfile
ARG PORT
ENTRYPOINT ["/bin/bash", "-c", "export PORT=8080; echo \"starting on port $PORT\"; someBinary"]
```
^^^ That will be very confusing, because the value for `$PORT` will be different in the `echo` statement than it is in `someBinary`. In the `echo` statement, it will be the earthly `ARG`; in someBinary, it will be `8080`.

**Comment by cirego** [comment]
I don't suppose there's any update on this? I'm trying to use arguments to define the name of a binary that should be the entrypoint for the image that I'm building -- and I'd love to avoid using the shell-form if possible. Thanks!
